### DISCUSSION TITLE
Field alias returning null while using the cacheExchange. I need some help

### DISCUSSION THREAD

**Discussion by edusig** [urql-graphql/urql#3134] [Q&A] [answered]
## The problem

I'm using alias for some image fields to generate thumbnails of them. The browser's network response shows every field with a value, however when I log the current query data the original field is `null`

## Context and Example

I'm using urql in a Nextjs project with a schema aware `cacheExchange` and I just updated the urql packages to the latest versions

Here is an example of a query:

```gql
query User {
  user {
    avatar {
      url(options: { size: "96x96", crop: "center center" })
      webp: url(options: { size: "96x96", format: WEBP, crop: "center center" })
      avif: url(options: { size: "96x96", format: AVIF, crop: "center center" })
      inlineBlur: url(
        options: { size: "32x32", format: WEBP, crop: "center center", inline: true, blur: 2 }
      )
    }
    thumbnail: avatar {
      url(options: { size: "32x32", crop: "center center" })
      webp: url(options: { size: "32x32", format: WEBP, crop: "center center" })
      avif: url(options: { size: "32x32", format: AVIF, crop: "center center" })
      inlineBlur: url(
        options: { size: "16x16", format: WEBP, crop: "center center", inline: true, blur: 2 }
      )
    }
  }
}
```

That query will return something like this:
```json
{
  "data": {
    "user": {
      "avatar": {
        "url": "https://example.com/example-image.png",
        "webp": "https://example.com/example-image.webp",
        "avif": "https://example.com/example-image.avif",
        "inlineBlur": "<BASE64_IMAGE>",
        "__typename": "ImageType"
      }
      "thumbnail": {
        "url": "https://example.com/example-thumbnail.png",
        "webp": "https://example.com/example-thumbnail.webp",
        "avif": "https://example.com/example-thumbnail.avif",
        "inlineBlur": "<BASE64_IMAGE>",
        "__typename": "ImageType"
      }
    }
  }
}
```

But logging the `data.user` object will output:
```ts
{
  avatar: null,
  thumbnail: {
    url: 'https://example.com/example-thumbnail.png',
    webp: 'https://example.com/example-thumbnail.webp',
    avif: 'https://example.com/example-thumbnail.avif',
    inlineBlur: '<BASE64_IMAGE>',
    __typename: 'ImageType',
  }
}
```

## Possible workarounds

Changing the query to the example below works, but I also generate images in some other formats and would be really nice to have a separate object and not having to alias every field
```gql
query User {
  user {
    avatar {
      url(options: { size: "96x96", crop: "center center" })
      webp: url(options: { size: "96x96", format: WEBP, crop: "center center" })
      avif: url(options: { size: "96x96", format: AVIF, crop: "center center" })
      inlineBlur: url(
        options: { size: "32x32", format: WEBP, crop: "center center", inline: true, blur: 2 }
      )
      thumbnailUrl: url(options: { size: "32x32", crop: "center center" })
      thumbnailWebp: url(options: { size: "32x32", format: WEBP, crop: "center center" })
      thumbnailAvif: url(options: { size: "32x32", format: AVIF, crop: "center center" })
      thumbnailInlineBlur: url(
        options: { size: "16x16", format: WEBP, crop: "center center", inline: true, blur: 2 }
      )
    }
  }
}
```

Removing the `cacheExchange` also makes it work, but I rely on the resolvers to transform some dates and numbers.

## Question 

Am I doing something wrong or missing something?

This problem is also making some queries trigger stale on each other indefinitely, but it could be that one of the queries has the alias and the other doesn't.

**Comment by kitten** [comment]
Hm, I think there might be a bug hidden here, but it's hard to tell without a reproduction or more details. Basically, it's hard to isolate this to any one cause, because without a reproduction, I can't really see what's a factor, and I also don't know if it's related to one of the following things that are in your example that may or may not be in your apps, like:
- `keys` generators
- key fields that are missing in the example above
- schema awareness could play a factor
- fragments or other indirection

Essentially, `ImageType` is a keyless type, which means it'll be embedded, and because its key is derived from the parent it'll be `[parentKey].avatar`, always, meaning your two selection sets will be combined regardless.

I don't quite know why one of them becomes `null` however, but again, hard to tell without looking at code; there may be a bug, or there may be a problem with something I'm not seeing here.

---

Unrelated Note: I'm putting an unrelated note here, because I do still want to find out where this bug is hiding, but I also want to comment on something I can't unsee 😅 

I'd recommend a small schema change. Of course, this may solve this problem, but again, I'm interested in finding this bug regardless.

`Avatar.ImageType` seems all good, but it seems kind of redundant, because the schema is doing "too much" of what the client doesn't have to know about. Now, capabilities aren't bad, but for instance, if you know that you always want to retrieve N images of a given format with the same size, then I find it a bit redundant to have `size: "32x32"` on `url` rather than on a hypothetical parent type 🤔 

---

Anyway, small side-note, I want to find a way to figure out what's going on here without spending too much time on needing you to reproduce it, if that's inconvenient, so to ease this process a bit:
- do you see any warnings in the console?
- are these queries exactly what's sent to the API (other selection sets outside of this one don't matter, probably)
- are you leaving out potential key fields?
- does the query eventually settle or is it always incomplete?
- what happens if you disable schema awareness by leaving out the `schema` option?
- what happens if you alias both `avatar` fields?
- what is `result.operation.context.meta.cacheOutcome` set to over time?

**Reply by edusig** [reply]
First off thanks for the quick reply and taking your time to help me. It took me some time to answer cause I was trying lots of variations on the queries, configs, page ssr/ssg, fragments and aliases.

I did simplify a lot the original query for the post, the `ImageType` does have an id and I'm actually using fragments to get every user field and the images have standardized sizes so there are a set of fragments on the Image type I will post here the full query to give more context (at the end).

I do have a fairly complex client config. My current config deals with the SSR cache, session cookies, a subscription exchange, a scalar serializer exchange (I modified the exchange from [this repo](https://github.com/marcinkoziej/urql-serialize-scalars-exchange) only to remove the deserialization part), a cache exchange with keys, the schema and as I've said earlier I rely on the resolvers (which I auto generate based on the field's Scalar types to convert dates to date objects and decimal (string) to floats). I will also put the configuration at the end.

--- 

I was originally going to post here last week about the queries triggering stale on each other. After updating "@urql/core" from version "3.0.5" to "3.2.1" there was a page (ongoing funding rounds page) that started having these indeterminate queries (Between a user query with a list of startups and a list of funding rounds query that selected fields from the same startup). I actually tried to reproduce it in this repo (https://github.com/edusig/urql-cache-exchange-example) using a public api but couldn't reproduce. Also remember that removing the schema option in the cacheExchange stopped that and also that page was using `getStaticProps` (from nextjs) and by changing it to a `getServerSideProps` (with the same queries inside) or removing it completely it also did not trigger the stale and the queries settled. 

Eventually I simplified some fragments, added a composite key to a type (the StartupUser type you can see below in the config)  and removed an alias for the startup logo in the funding rounds list and it seemed to have stopped the problem. It was last Friday so I might have gave up or imagined that it worked, cause now that I'm loading that exact page right now and it has the same problem.

To give you more context this website is for our startup investment platform and we are changing the API and redesigning some pages so I took the opportunity to migrate from Apollo to Urql (I have recently done this in another project and had no problems so most of the configuration was copied from there, note that this other project is using "@urql/core@3.0.5")

--- 

About the ImageType unrelated note: our backend is a Python + Django + [Strawberry](https://strawberry.rocks/) application. And we are using  [Sorl-thumbnail](https://pypi.org/project/sorl-thumbnail/11.12/) to give  the frontend more flexibility to specify the size and format of the images we generate and cache on the fly. It is by all means inefficient and was made like that so we could try out new formats, image crop, inline blur, etc without having to change the API. Maybe it is time to decide on a final structure and remove the frontend control of the options XD.

--- 

Important things to notice:
- Removing the image aliases from both queries makes them settle normally (even when loading queries on SSR or SSG)
- I'm a little short on time to launch this website update so I will probably just remove the aliases for now. I will try to reproduce this problem in a public repo if I have some time (I will also talk with my co-worker that did the backend for this app to see if we can make a public dummy project with similar structures/queries). If there are any more things I can try out or information to help understand the problem feel free to ask, I also would like to understand what is causing it.
- The `StartupUser` type caught my attention by being the only keyless type between these 2 queries. I will see if we can change it to be just a list of `Startups`.

Answering the side note questions:

> do you see any warnings in the console?

I did resolve all the warnings, mostly for missing key fields (You can see on the config at the end). There are no other warnings in the console


> are these queries exactly what's sent to the API (other selection sets outside of this one don't matter, probably)

The full query as it was in the request sent to the API will be at the end XD


> are you leaving out potential key fields?

The Image type does have a unique ID, and I did try adding it to the query, tried removing it and mapping the keys on the cacheExchange to the url, tried adding it and also setting the key to the id + url and there was no change.


> does the query eventually settle or is it always incomplete?

The way it is now (after updating the urql packages to the latest version) in development, the ongoing funding rounds page will only settle the 2 queries (current user and latest funding rounds) if I use `getStaticProps` and make it empty (don't query the list of ongoing funding rounds on the server, the current user cannot be queried anyway since in a server generated static page there will be no session).


> what happens if you disable schema awareness by leaving out the schema option?

The whole user data and funding rounds data is null unless I also remove the keys and resolvers options. Also tried changing the page from `getStaticProps` to `getServerSideProps` or removing the queries inside both and nothing changed. Tried loading other pages and seems like every query is returning null inside data.


> what happens if you alias both avatar fields?

The one that comes the last on the query/fragment gets the data with or without alias. Some examples:

```gql
fragment userImages on User {
  avatar {
    ...avatarImages
  }
  thumbnail: avatar { # <- this one gets the data
    ...avatarThumbnailImages
  }
}

fragment userImages2 on User {
  thumbnail: avatar {
    ...avatarThumbnailImages
  }
  avatar { # <- this one gets the data
    ...avatarImages
  }
}

fragment userImages3 on User {
  otherAvatar: avatar {
    ...avatarImages
  }
  thumbnail: avatar { # <- this one gets the data
    ...avatarThumbnailImages
  }
}

fragment userImages4 on User {
  thumbnail: avatar {
    ...avatarThumbnailImages
  }
  otherAvatar: avatar { # <- this one gets the data
    ...avatarImages
  }
}
```
> what is result.operation.context.meta.cacheOutcome set to over time?

After a couple of seconds it looks like this:

![Screenshot 2023-04-06 at 1 12 50 AM](https://user-images.githubusercontent.com/810728/230270589-83330a20-caae-4f68-9273-14c9eebb610a.png)

--- 
Here is the full user query including a list of its startups and the full funding rounds query with some startup information:

```gql
query Me {
  me {
    id
    ...userFields
    ...userDetailFields
    startups {
      user {
        id
        __typename
      }
      startup {
        id
        name
        __typename
      }
      __typename
    }
    __typename
  }
}
fragment userFields on User {
  emails {
    id
    email
    verified
    primary
    __typename
  }
  name
  ...userImages
  __typename
}
fragment userImages on User {
  avatar {
    ...avatarImages
    __typename
  }
  thumbnail: avatar {
    ...avatarThumbnailImages
    __typename
  }
  __typename
}
fragment avatarImages on ImageType {
  id
  name
  url(options: { size: "96x96", crop: "center center" })
  webp: url(options: { size: "96x96", format: WEBP, crop: "center center" })
  avif: url(options: { size: "96x96", format: AVIF, crop: "center center" })
  inlineBlur: url(
    options: { size: "32x32", format: WEBP, crop: "center center", inline: true, blur: 2 }
  )
  __typename
}
fragment avatarThumbnailImages on ImageType {
  id
  name
  url(options: { size: "32x32", crop: "center center" })
  webp: url(options: { size: "32x32", format: WEBP, crop: "center center" })
  avif: url(options: { size: "32x32", format: AVIF, crop: "center center" })
  inlineBlur: url(
    options: { size: "16x16", format: WEBP, crop: "center center", inline: true, blur: 2 }
  )
  __typename
}
fragment userDetailFields on User {
  doc
  gender
  civilState
  investmentProfile
  isPep
  externalInvestment
  totalInvestment
  otherInvestments
  visibility
  phone
  whatsapp
  birthday
  occupation
  postalCode
  street
  streetNumber
  district
  complement
  city
  state
  annualGrossIncome
  origin
  __typename
}

query FundingConn(
  $first: Int = 100
  $after: String
  $filters: FundingFilter
  $order: FundingOrder
) {
  fundingConn(first: $first, after: $after, filters: $filters, order: $order) {
    pageInfo {
      hasNextPage
      hasPreviousPage
      startCursor
      endCursor
      __typename
    }
    totalCount
    edges {
      node {
        ...fundingSummaryFields
        __typename
      }
      __typename
    }
    __typename
  }
}
fragment fundingSummaryFields on Funding {
  id
  slug
  headline
  status
  confirmedAmount
  targetAmount
  totalAmount
  minPercent
  startup {
    id
    name
    doc
    homepage
    headline
    ...startupImages
    __typename
  }
  closureDate
  finishedAt
  __typename
}
fragment startupImages on Startup {
  logo {
    ...avatarBannerImage
    __typename
  }
  thumbnail: logo {
    ...avatarStartupThumbnailImages
    __typename
  }
  __typename
}
fragment avatarBannerImage on ImageType {
  id
  name
  url(options: { size: "352x120" })
  webp: url(options: { size: "352x120", format: WEBP })
  avif: url(options: { size: "352x120", format: AVIF })
  inlineBlur: url(options: { size: "48x32", format: WEBP, inline: true, blur: 2 })
  __typename
}
fragment avatarStartupThumbnailImages on ImageType {
  id
  name
  url(options: { size: "32x32", crop: "center center" })
  webp: url(options: { size: "32x32", format: WEBP, crop: "center center" })
  avif: url(options: { size: "32x32", format: AVIF, crop: "center center" })
  inlineBlur: url(
    options: { size: "16x16", format: WEBP, crop: "center center", inline: true, blur: 2 }
  )
  __typename
}
```

Here is my full config and how I use it:

```ts
// client.ts file
export const getGraphQLClientConfig = ({ cookie, ssrCache }: GetGraphQLClientOptions) => {
  const url = getAPIEndpoint();
  const wsClient =
    typeof window !== 'undefined'
      ? createWSClient({
          url: getAPIEndpoint().replace('http', 'ws'),
        })
      : null;
  const config: ClientOptions = {
    url,
    fetchOptions:
      cookie != null
        ? {
            headers: { cookie },
          }
        : undefined,
    exchanges: [
      cacheExchange({
        // schema: schema as IntrospectionData,
        schema: schema as any,
        keys: {
          ImageType: data => data.url?.toString() ?? null,
          StartupUser: data => `${(data.startup as Startup).id}${(data.user as User).id}`,
          FundingLimits: () => null,
        },
        resolvers,
      }),
      ...(wsClient != null
        ? [
            subscriptionExchange({
              forwardSubscription: request => ({
                subscribe: sink => {
                  return {
                    unsubscribe: wsClient.subscribe(
                      { ...request, query: request.query ?? '' },
                      sink,
                    ),
                  };
                },
              }),
            }),
          ]
        : []),
      ...(ssrCache != null ? [ssrCache] : []),
      createSerializeScalarsExchange(scalarLocations),
      fetchExchange,
    ] as any[],
  };
  return config;
};

export const getGraphQLClient = (options: GetGraphQLClientOptions) => {
  const isSSR = typeof window === 'undefined';
  if (client != null && !isSSR) return client;
  const config = getGraphQLClientConfig(options);
  client = isSSR ? initUrqlClient(config, false) : createClient(config);
  return client;
};

// with-graphql-client.ts file
export const withGraphQLClientHOC = withUrqlClient((ssrCache, ctx) =>
  getGraphQLClientConfig({ ssrCache, cookie: ctx?.req?.headers.cookie }),
);

// _app.ts file
export default withGraphQLClientHOC(CustomApp);
```

A part of our schema.gql
```gql
type ImageType implements Node {
  id: GlobalID!
  name: String!
  url(options: ImageProcessingOptions = null): String!
}

"""Image processing options."""
input ImageProcessingOptions {
  """The size to resize the image as wxh (e.g. '200x200')"""
  size: String!

  """If we should return the image inlined (i.e. as base64)."""
  inline: Boolean! = false

  """If we should add a padding to the image after resizing it."""
  padding: Boolean! = false

  """Crop options for the image"""
  crop: String = null

  """If we should add a padding to the image after resizing it."""
  paddingColor: String! = "#ffffff"

  """The quality to use when converting (0-100)."""
  quality: Int! = 95

  """The colorspace for the resulting thumbnail."""
  colorspace: ImageColorspace! = RGB

  """The resulting file format. If null, the format will be preserved"""
  format: ImageFormat = null

  """Gaussian blur radius."""
  blur: Int! = 0
}

type User implements Node & Named & WithAvatar {
  id: GlobalID!
  name: String!
  avatar: ImageType
  username: String!
  firstName: String!
  lastName: String!
  phone: PhoneNumber
  whatsapp: PhoneNumber
  gender: UserGender
  createdAt: DateTime!
  updatedAt: DateTime!
  birthday: Date
  lastLogin: DateTime
  dateJoined: DateTime!
  origin: UserOrigin
  civilState: UserCivilState
  investmentProfile: UserInvestmentProfile
  isPep: Boolean
  annualGrossIncome: Decimal
  internalInvestment: Decimal!
  otherInvestments: Decimal
  externalInvestment: Decimal
  totalInvestment: Decimal!
  visibility: Boolean!
  occupation: String
  postalCode: String
  street: String
  streetNumber: String
  district: String
  complement: String
  city: String
  state: String
  doc(formatted: Boolean! = false): String
  emails: [UserEmailAddress!]!
  socialAccounts: [SocialAccount!]!
  startups: [StartupUser!]!
}

type StartupUser {
  user: User!
  startup: Startup!
  isOwner: Boolean!
  isAdmin: Boolean!
}

type Startup implements Node & Named {
  id: GlobalID!
  name: String!
  email: String
  legalName: String
  headline: String!
  description: String!
  homepage: String
  doc: String
  phone: PhoneNumber
  whatsapp: PhoneNumber
  foundedAt: Date
  postalCode: String
  street: String
  streetNumber: String
  district: String
  complement: String
  city: String
  state: String
  revenueModel: RevenueModel
  audienceKind: AudienceKind
  fundings(filters: FundingFilter, order: FundingOrder): [Funding!]!
  logo: ImageType
  users: [StartupUser!]!
}
```

---

A somewhat related question:
- Is there a better way to serialize/deserialize complex types/objects than what I'm doing right now with the resolvers on the cacheExchange and the custom serialize scalars exchange?

**Reply by kitten** [reply, marked as answer]
I think I'm seeing the issue;
while you've got fragments like these for `ImageType`:

```
fragment avatarBannerImage on ImageType {
  id
  name
  url(options: { size: "352x120" })
  webp: url(options: { size: "352x120", format: WEBP })
  avif: url(options: { size: "352x120", format: AVIF })
  inlineBlur: url(options: { size: "48x32", format: WEBP, inline: true, blur: 2 })
  __typename
}
```

The keys configuration says:

```js
ImageType: data => data.url?.toString() ?? null,
```

In other words, you're using the `url` property of the `ImageType` objects, rather than the `id` field.
So, what happens is:

- response comes back
- the first `ImageType` selection set is written
  - Key A: `url` key is used or `null` is used
- the second `ImageType` selection set is written
  - Key B: `url` key is used or `null` is used (**NOTE:** this may be different)

So, at this point all data is written, but potentially to two different keys.
However, the `avatar` field is always the same, so now, `avatar` is set to the second key only!

- response is queried from cache
- the first selection set is read from cache
  - Key B is used, but Key A data isn't found
- the second selection set is read
  - Key B data is found

So, in other words, if you were to use `null` always, you'd be fine,
if you were to use `id` and always queried that field and used `ImageType: data => data.id`, you'd also be fine.

However, there's an important note here:
- If the keys differ on the same field of the parent, we don't have a warning for that and data will go missing
- If the `keys.ImageType` function returns `null`, i.e. no `url` field is queried, the warning is silenced because it's not returning `undefined`

So, you can either configure:

```
ImageType: data => data.id,
```

And make sure you also always query `id`.
Or, if the `id` is rather irrelevant because the `ImageType` per parent doesn't change, you can also do:

```
ImageType: () => null,
```

and both selection sets will be combined as you expect.

**Reply by kitten** [reply]
So basically, so this doesn't get lost, I'd also change this:
```
StartupUser: data => `${(data.startup as Startup).id}${(data.user as User).id}`,
```
to this:
```
StartupUser: data => data.startup && data.user
  ? `${(data.startup as Startup).id}${(data.user as User).id}`
  : undefined,
```

so that you can see if you have missing key fields using the warning in the console, to protect you from missing key fields. That's because we currently don't have tooling other than these warnings to protect you from forgetting to use the key fields

**Reply by edusig** [reply]
That did the trick XD.

So just to see if I got this, the original field and the alias should either have the same unique `id` or `null`.
And by mapping to use the url (which would be different for the alias) the second field overwrote the first one and the data was lost, therefore the first field data would be `null` on the cache.

About your second reply the `KeyGenerator` type allows the function to only return `string | null` so returning `undefined` would raise an error in Typescript. Would returning `null` show the warning, or `undefined` have a different behavior?

**Reply by kitten** [reply]
Re. the type error: yea, I just realised the type is incorrect and we should probably add an explicit entry in the documentation and a variation in the warning for it. This way we can document that Graphcache can still warn you if you return `undefined`.

The reasoning here is the same for resolvers, basically, you can always return `null` to say "this is the value `null`", and you can return `undefined` to say "this is a missing value". So for the keys, internally, it doesn't matter to Graphcache whether there's actually a `keys` config or not, `undefined` means it's missing, `null` means it's not missing and the entity becomes embedded.

Anyway, re. the original issue, yes, the problem here is that you have two identical "paths" to different entities, so they overwrite each other.

To explain it the other way around, if you were to write another query that _didn't_ use this aliased field, you'd still expect it to find the item. So, for instance, `{ alias: book(id: $id) { id } }` should resolve, even if another query did `{ book(id: $id) { id } }`, because conceptually, that's the same data.

In your case `avatar { url(options: { size: "352x120" }) }` and `alias: avatar { url(options: { size: "352x120" }) }` point to the same data in that same manner, however, if the keys start differing, then you basically "split" the data.

I'm still thinking about how we could handle this more gracefully however and warn about these kinds of issues instead of failing in this non-obvious way

**Reply by edusig** [reply]
Well thanks again for the help. The warning, fix on the return type and mention on the documentation will help a lot XD
